### ISSUE TITLE
not able to run indexer for svn due to "Server SSL certificate verification failed: issuer is not trusted" on docker container

### ISSUE THREAD

**Issue by rmshnair** [oracle/opengrok#3218]
```
opengrok    | 10-Sep-2020 03:57:38.611 SEVERE [invalidate-repos-15] org.opengrok.indexer.util.Executor.exec Failed to read from process: svn
opengrok    |    java.io.IOException: An error occurred while parsing the xml output
opengrok    |            at org.opengrok.indexer.history.SubversionHistoryParser.processStream(SubversionHistoryParser.java:225)
opengrok    |            at org.opengrok.indexer.util.Executor.exec(Executor.java:230)
opengrok    |            at org.opengrok.indexer.history.SubversionHistoryParser.parse(SubversionHistoryParser.java:196)
opengrok    |            at org.opengrok.indexer.history.SubversionRepository.getHistory(SubversionRepository.java:292)
opengrok    |            at org.opengrok.indexer.history.SubversionRepository.determineCurrentVersion(SubversionRepository.java:412)
opengrok    |            at org.opengrok.indexer.history.RepositoryFactory.getRepository(RepositoryFactory.java:234)
opengrok    |            at org.opengrok.indexer.history.RepositoryFactory.getRepository(RepositoryFactory.java:151)
opengrok    |            at org.opengrok.indexer.history.RepositoryFactory.getRepository(RepositoryFactory.java:274)
opengrok    |            at org.opengrok.indexer.history.HistoryGuru$3.run(HistoryGuru.java:877)
opengrok    |            at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
opengrok    |            at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
opengrok    |            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
opengrok    |            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
opengrok    |            at java.base/java.lang.Thread.run(Thread.java:834)
opengrok    |    Caused by: org.xml.sax.SAXParseException; lineNumber: 3; columnNumber: 1; XML document structures must start and end within the same entity.
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1471)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.endEntity(XMLDocumentFragmentScannerImpl.java:933)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.endEntity(XMLDocumentScannerImpl.java:562)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.endEntity(XMLEntityManager.java:1504)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.load(XMLEntityScanner.java:1925)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.skipChar(XMLEntityScanner.java:1560)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2746)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:534)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:888)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:824)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1216)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:635)
opengrok    |            at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:324)
opengrok    |            at java.xml/javax.xml.parsers.SAXParser.parse(SAXParser.java:197)
opengrok    |            at org.opengrok.indexer.history.SubversionHistoryParser.processStream(SubversionHistoryParser.java:223)
opengrok    |            ... 13 more
opengrok    | 10-Sep-2020 03:57:38.612 WARNING [invalidate-repos-15] org.opengrok.indexer.util.Executor.exec Non-zero exit status -1 from command [svn log --non-interactive --username admin --password admin --xml -v -l1 $'/opengrok/src/resources@'] in directory /opengrok/src/resources



/opengrok/src/resources# svn log --non-interactive --username admin --password admin --xml -v -l1 $'/opengrok/src/resources@'
<?xml version="1.0" encoding="UTF-8"?>
<log>
svn: E170013: Unable to connect to a repository at URL 'https://ar-shr-d-spk-03.cs.boeing.com/repos/resources'
svn: E230001: Server SSL certificate verification failed: issuer is not trusted
```

**Describe the bug**
since --trust-server-cert is not available as part of the svn command, indexer is failing to retrieve the history of svn projects.

**To Reproduce**
configure a svn source and start the container

**Expected behavior**
execute all svn commands with --trust-server-cert option.

**Comment by vladak** [comment]
Subversion uses OpenSSL (`libssl.so`) for TLS I believe. The official OpenGrok docker image is based on the `tomcat:9-jdk11` image that in turn is based on Debian stretch (https://microbadger.com/images/tomcat:9-jdk11). This is just a bare image that does not have OpenSSL - it probably gets installed as a dependency of one of the packages which are installed in one of the opengrok image layers.

The https://packages.debian.org/stretch/ca-certificates package can be added and would install trust anchor for all apps using OpenSSL however I am not sure if that would help in this particular case - are you perhaps using custom CA ? (i.e. the one not on the list of Mozilla's certificates shipped in the package)

**Comment by rmshnair** [comment]
as part of debugging procedure, i found ca-certificates are already present in the image. may be as you said, it is installed as dependency. i tried installing the same ca and root chain, ran update-ca-certificates,  however still the error persists.
i have also updated in ~/.subversion/servers with "ssl-trust-default-ca = yes". ssl-authority-files with actual ca files, but none worked.

two methods, i could successfully start the indexer is, 
1. get into to the container and run the svn log command removing --non-interactive which will ask for accept certificate and provide "p" to accept permenantly.
2. the alternative way that works is provide both "--non-interactive" and "--trust-server-cert" options to svn command, which will automatically accept temporarily on every time.

looks like we are missing this flag in the command builder. or suggest me an work around.

**Comment by vladak** [comment]
To clarify: is the CA certificate used by the Subversion server in question part of the `/etc/ssl/certs/` as installed by the `ca-certificates` package or not ?

What happens if you connect to the Subversion server with `openssl s_client` running from inside the Docker container ?

The `--trust-server-cert` flag should not be used by default in the history log executor. It should only be used for testing purposes as it creates a security problem.

**Comment by vladak** [comment]
In fact, the option was removed in #1564

**Comment by vladak** [comment]
At the risk of providing Debian support here:

TL;DR: The `update-ca-certificates` script responsible for adding the hashed symlinks to `/etc/ssl/certs` directory is quite picky about locations and file names. Make sure that it said it actually added the certificates. Successful run should state 1 new certificate was added.

Here's my test:

First, I created self-signed certificate and associated RSA key par in `/var/tmp` and started TLS server on port 4433 using the generated data:
```
root@d8d74441014b:/var/tmp#  openssl req   -x509 -nodes -days 365   \
>      -subj '/C=CZ/ST=Czech republic/L=Prague/CN=foo.com' \
>      -newkey rsa:2048 -keyout mykey.pem -out mycert.pem
Generating a RSA private key
..........................................................+++++
.....................................+++++
writing new private key to 'mykey.pem'
-----
root@d8d74441014b:/var/tmp# openssl s_server -cert mycert.pem -key mykey.pem -www -port 4433
Using default temp DH parameters
ACCEPT
```
Tried connecting with `s_client`:
```
root@d8d74441014b:/etc/ssl/certs# openssl s_client -connect localhost:4433
CONNECTED(00000003)
Can't use SSL_get_servername
depth=0 C = CZ, ST = Czech republic, L = Prague, CN = foo.com
verify error:num=18:self signed certificate
verify return:1
depth=0 C = CZ, ST = Czech republic, L = Prague, CN = foo.com
verify return:1
---
Certificate chain
 0 s:C = CZ, ST = Czech republic, L = Prague, CN = foo.com
   i:C = CZ, ST = Czech republic, L = Prague, CN = foo.com
---


...

---
read R BLOCK
^C
```
Notice the `verify error:num=18:self signed certificate` error message at the beginning.

Then I copied the certificate to the `/usr/local/share/ca-certificates/` directory. Now, the `update-ca-certificates` script is quite picky about locations and naming. I had to copy the `/var/tmp/mycert.pem` file to  `/usr/local/share/ca-certificates/mycert.crt` for this to work.

Next I ran the script:
```
root@d8d74441014b:/etc/ssl/certs# update-ca-certificates 
Updating certificates in /etc/ssl/certs...
1 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
```
Yep, added new certificate:
```
root@d8d74441014b:/etc/ssl/certs# ls -l | grep mycert
lrwxrwxrwx 1 root root     10 Sep 10 20:04 612da7de.0 -> mycert.pem
lrwxrwxrwx 1 root root     43 Sep 10 20:04 mycert.pem -> /usr/local/share/ca-certificates/mycert.crt
```

I rerun the client, this time it did not complain about self-signed certificate:
```
root@d8d74441014b:/etc/ssl/certs# openssl s_client -connect localhost:4433
CONNECTED(00000003)
Can't use SSL_get_servername
depth=0 C = CZ, ST = Czech republic, L = Prague, CN = foo.com
verify return:1
---
Certificate chain
 0 s:C = CZ, ST = Czech republic, L = Prague, CN = foo.com
   i:C = CZ, ST = Czech republic, L = Prague, CN = foo.com
```

**Comment by rmshnair** [comment]
i think your point helps, however, i need to add explicity ca certs into .subversion/servers file so that svn can consume to validate those. Thanks.
